File: python_backend/services/security_gateway.py
"""
SecurityGateway - Backend Security Gateway

Provides input validation, sanitization, and Arabic error message translation
with graceful degradation for invalid inputs.

Week 2 Task 2.1: Security Implementation
"""

import re
import json
from typing import Any, Dict, Optional, Union, List
from datetime import datetime
from enum import Enum

from fastapi import Request, HTTPException, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field, validator

from core.security_logger import SecurityLogger, SecurityEvent, SecurityEventType


class SecurityErrorSeverity(str, Enum):
    """Security error severity levels"""
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"


class SecurityError(BaseModel):
    """Security error model with Arabic translation"""
    code: str
    message: str
    message_ar: str  # Arabic translation
    severity: SecurityErrorSeverity = SecurityEventType.ERROR
    field: Optional[str] = None
    details: Optional[Dict[str, Any]] = None


class ValidationResult(BaseModel):
    """Validation result model"""
    valid: bool
    sanitized: Optional[Any] = None
    error: Optional[SecurityError] = None


class SecurityGateway:
    """Backend security gateway for input validation and sanitization"""
    
    # Dangerous patterns for XSS detection
    XSS_PATTERNS = [
        re.compile(r'<script', re.IGNORECASE),
        re.compile(r'javascript:', re.IGNORECASE),
        re.compile(r'onerror=', re.IGNORECASE),
        re.compile(r'onload=', re.IGNORECASE),
        re.compile(r'eval\(', re.IGNORECASE),
        re.compile(r'expression\(', re.IGNORECASE),
        re.compile(r'data:text/html', re.IGNORECASE),
    ]
    
    # SQL injection patterns
    SQL_PATTERNS = [
        re.compile(r'\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE)\b', re.IGNORECASE),
        re.compile(r"('|(\\')|(;)|(\\)|(\/\*)|(\*\/)|(--)|(#))"),
    ]
    
    def __init__(self):
        self.security_logger = SecurityLogger()
        self._error_messages = self._load_error_messages()
    
    def _load_error_messages(self) -> Dict[str, Dict[str, str]]:
        """Load error messages from locale files"""
        messages = {
            'INVALID_INPUT': {
                'en': 'Input is required',
                'ar': 'المدخل مطلوب'
            },
            'XSS_ATTEMPT': {
                'en': 'Input contains potentially dangerous content',
                'ar': 'المدخل يحتوي على محتوى خطير محتمل'
            },
            'SQL_INJECTION_ATTEMPT': {
                'en': 'Input contains potentially dangerous SQL patterns',
                'ar': 'المدخل يحتوي على أنماط SQL خطيرة محتملة'
            },
            'MIN_LENGTH': {
                'en': 'Input must be at least {min} characters',
                'ar': 'يجب أن يكون المدخل {min} أحرف على الأقل'
            },
            'MAX_LENGTH': {
                'en': 'Input must be no more than {max} characters',
                'ar': 'يجب ألا يتجاوز المدخل {max} حرف'
            },
            'INVALID_NUMBER': {
                'en': 'Input must be a valid number',
                'ar': 'يجب أن يكون المدخل رقماً صحيحاً'
            },
            'VALIDATION_ERROR': {
                'en': 'Validation failed: {error}',
                'ar': 'فشل التحقق: {error}'
            },
        }
        return messages
    
    def get_error_message(self, code: str, language: str = 'en', **kwargs) -> str:
        """Get localized error message"""
        messages = self._error_messages.get(code, {})
        template = messages.get(language, messages.get('en', code))
        
        try:
            return template.format(**kwargs)
        except (KeyError, ValueError):
            return template
    
    def validate_input(
        self,
        input_data: Any,
        min_length: Optional[int] = None,
        max_length: Optional[int] = None,
        min_value: Optional[float] = None,
        max_value: Optional[float] = None,
        pattern: Optional[re.Pattern] = None,
        required: bool = True
    ) -> ValidationResult:
        """
        Validate and sanitize input data
        
        Args:
            input_data: Input to validate
            min_length: Minimum length for strings
            max_length: Maximum length for strings
            min_value: Minimum value for numbers
            max_value: Maximum value for numbers
            pattern: Regex pattern for string validation
            required: Whether input is required
        
        Returns:
            ValidationResult with validation status and sanitized data
        """
        try:
            # Check if required
            if required and (input_data is None or input_data == ''):
                return ValidationResult(
                    valid=False,
                    error=SecurityError(
                        code='INVALID_INPUT',
                        message=self.get_error_message('INVALID_INPUT', 'en'),
                        message_ar=self.get_error_message('INVALID_INPUT', 'ar'),
                        severity=SecurityErrorSeverity.ERROR
                    )
                )
            
            # String validation
            if isinstance(input_data, str):
                return self._validate_string(
                    input_data, min_length, max_length, pattern
                )
            
            # Number validation
            if isinstance(input_data, (int, float)):
                return self._validate_number(
                    input_data, min_value, max_value
                )
            
            # List validation
            if isinstance(input_data, list):
                return self._validate_list(input_data)
            
            # Dict validation
            if isinstance(input_data, dict):
                return self._validate_dict(input_data)
            
            # Default: valid
            return ValidationResult(valid=True, sanitized=input_data)
            
        except Exception as e:
            error_msg = str(e)
            return ValidationResult(
                valid=False,
                error=SecurityError(
                    code='VALIDATION_ERROR',
                    message=self.get_error_message('VALIDATION_ERROR', 'en', error=error_msg),
                    message_ar=self.get_error_message('VALIDATION_ERROR', 'ar', error=error_msg),
                    severity=SecurityErrorSeverity.ERROR
                )
            )
    
    def _validate_string(
        self,
        input_str: str,
        min_length: Optional[int] = None,
        max_length: Optional[int] = None,
        pattern: Optional[re.Pattern] = None
    ) -> ValidationResult:
        """Validate string input"""
        # Check for XSS patterns
        for pattern_check in self.XSS_PATTERNS:
            if pattern_check.search(input_str):
                self._log_security_event(
                    SecurityEventType.SUSPICIOUS_REQUEST,
                    'XSS_ATTEMPT',
                    {'input_preview': input_str[:100]}
                )
                return ValidationResult(
                    valid=False,
                    error=SecurityError(
                        code='XSS_ATTEMPT',
                        message=self.get_error_message('XSS_ATTEMPT', 'en'),
                        message_ar=self.get_error_message('XSS_ATTEMPT', 'ar'),
                        severity=SecurityErrorSeverity.ERROR
                    )
                )
        
        # Check for SQL injection patterns
        for sql_pattern in self.SQL_PATTERNS:
            if sql_pattern.search(input_str):
                self._log_security_event(
                    SecurityEventType.SUSPICIOUS_REQUEST,
                    'SQL_INJECTION_ATTEMPT',
                    {'input_preview': input_str[:100]}
                )
                return ValidationResult(
                    valid=False,
                    error=SecurityError(
                        code='SQL_INJECTION_ATTEMPT',
                        message=self.get_error_message('SQL_INJECTION_ATTEMPT', 'en'),
                        message_ar=self.get_error_message('SQL_INJECTION_ATTEMPT', 'ar'),
                        severity=SecurityErrorSeverity.ERROR
                    )
                )
        
        # Check length constraints
        if min_length is not None and len(input_str) < min_length:
            return ValidationResult(
                valid=False,
                error=SecurityError(
                    code='MIN_LENGTH',
                    message=self.get_error_message('MIN_LENGTH', 'en', min=min_length),
                    message_ar=self.get_error_message('MIN_LENGTH', 'ar', min=min_length),
                    severity=SecurityErrorSeverity.ERROR
                )
            )
        
        if max_length is not None and len(input_str) > max_length:
            return ValidationResult(
                valid=False,
                error=SecurityError(
                    code='MAX_LENGTH',
                    message=self.get_error_message('MAX_LENGTH', 'en', max=max_length),
                    message_ar=self.get_error_message('MAX_LENGTH', 'ar', max=max_length),
                    severity=SecurityErrorSeverity.ERROR
                )
            )
        
        # Check pattern
        if pattern is not None and not pattern.match(input_str):
            return ValidationResult(
                valid=False,
                error=SecurityError(
                    code='PATTERN_MISMATCH',
                    message='Input does not match required pattern',
                    message_ar='المدخل لا يطابق النمط المطلوب',
                    severity=SecurityErrorSeverity.ERROR
                )
            )
        
        # Sanitize
        sanitized = self._sanitize_string(input_str)
        
        return ValidationResult(valid=True, sanitized=sanitized)
    
    def _validate_number(
        self,
        input_num: Union[int, float],
        min_value: Optional[float] = None,
        max_value: Optional[float] = None
    ) -> ValidationResult:
        """Validate number input"""
        if not isinstance(input_num, (int, float)) or not (isinstance(input_num, float) and input_num.is_integer() or isinstance(input_num, int)):
            if isinstance(input_num, float) and input_num.is_integer():
                input_num = int(input_num)
        
        if min_value is not None and input_num < min_value:
            return ValidationResult(
                valid=False,
                error=SecurityError(
                    code='MIN_VALUE',
                    message=f'Input must be at least {min_value}',
                    message_ar=f'يجب أن يكون المدخل {min_value} على الأقل',
                    severity=SecurityErrorSeverity.ERROR
                )
            )
        
        if max_value is not None and input_num > max_value:
            return ValidationResult(
                valid=False,
                error=SecurityError(
                    code='MAX_VALUE',
                    message=f'Input must be no more than {max_value}',
                    message_ar=f'يجب ألا يتجاوز المدخل {max_value}',
                    severity=SecurityErrorSeverity.ERROR
                )
            )
        
        return ValidationResult(valid=True, sanitized=input_num)
    
    def _validate_list(self, input_list: List[Any]) -> ValidationResult:
        """Validate list input"""
        sanitized = []
        errors = []
        
        for i, item in enumerate(input_list):
            result = self.validate_input(item, required=False)
            if result.valid:
                sanitized.append(result.sanitized)
            elif result.error:
                errors.append({
                    'index': i,
                    'error': result.error.dict()
                })
        
        if errors:
            return ValidationResult(
                valid=False,
                error=SecurityError(
                    code='ARRAY_VALIDATION_FAILED',
                    message='Array validation failed',
                    message_ar='فشل التحقق من المصفوفة',
                    severity=SecurityErrorSeverity.ERROR,
                    details={'errors': errors}
                )
            )
        
        return ValidationResult(valid=True, sanitized=sanitized)
    
    def _validate_dict(self, input_dict: Dict[str, Any]) -> ValidationResult:
        """Validate dictionary input"""
        sanitized = {}
        errors = []
        
        for key, value in input_dict.items():
            # Validate key
            key_result = self.validate_input(key, max_length=100)
            if not key_result.valid:
                errors.append({
                    'field': f'key:{key}',
                    'error': key_result.error.dict() if key_result.error else {}
                })
                continue
            
            # Validate value
            value_result = self.validate_input(value, required=False)
            if value_result.valid:
                sanitized[key_result.sanitized] = value_result.sanitized
            elif value_result.error:
                errors.append({
                    'field': key,
                    'error': value_result.error.dict()
                })
        
        if errors:
            return ValidationResult(
                valid=False,
                error=SecurityError(
                    code='OBJECT_VALIDATION_FAILED',
                    message='Object validation failed',
                    message_ar='فشل التحقق من الكائن',
                    severity=SecurityErrorSeverity.ERROR,
                    details={'errors': errors}
                )
            )
        
        return ValidationResult(valid=True, sanitized=sanitized)
    
    def _sanitize_string(self, input_str: str) -> str:
        """Sanitize string input"""
        # Remove dangerous characters
        sanitized = input_str
        sanitized = re.sub(r'[<>]', '', sanitized)  # Remove < and >
        sanitized = re.sub(r'javascript:', '', sanitized, flags=re.IGNORECASE)
        sanitized = re.sub(r'on\w+=', '', sanitized, flags=re.IGNORECASE)  # Remove event handlers
        sanitized = sanitized.strip()
        
        return sanitized
    
    def _log_security_event(
        self,
        event_type: SecurityEventType,
        error_code: str,
        details: Dict[str, Any]
    ):
        """Log security event"""
        event = SecurityEvent(
            event_type=event_type,
            timestamp=datetime.utcnow(),
            user_id=None,  # Will be set by caller if available
            ip_address=None,  # Will be set by caller if available
            user_agent=None,  # Will be set by caller if available
            details={
                'error_code': error_code,
                **details
            },
            severity='WARNING' if event_type == SecurityEventType.SUSPICIOUS_REQUEST else 'ERROR'
        )
        self.security_logger.log_event(event)
    
    def create_error_response(
        self,
        error: SecurityError,
        request: Optional[Request] = None,
        status_code: int = status.HTTP_400_BAD_REQUEST
    ) -> JSONResponse:
        """Create standardized error response with Arabic support"""
        # Detect language from request
        language = 'en'
        if request:
            accept_language = request.headers.get('Accept-Language', 'en')
            if 'ar' in accept_language.lower():
                language = 'ar'
        
        # Select message based on language
        message = error.message_ar if language == 'ar' else error.message
        
        response_data = {
            "error": {
                "code": error.code,
                "message": message,
                "message_ar": error.message_ar,
                "message_en": error.message,
                "severity": error.severity.value,
                "timestamp": datetime.utcnow().isoformat() + "Z",
            }
        }
        
        if error.field:
            response_data["error"]["field"] = error.field
        
        if error.details:
            response_data["error"]["details"] = error.details
        
        if request:
            response_data["error"]["path"] = str(request.url.path)
            response_data["error"]["method"] = request.method
        
        return JSONResponse(
            status_code=status_code,
            content=response_data
        )


# Global instance
_security_gateway_instance: Optional[SecurityGateway] = None


def get_security_gateway() -> SecurityGateway:
    """Get global SecurityGateway instance"""
    global _security_gateway_instance
    if _security_gateway_instance is None:
        _security_gateway_instance = SecurityGateway()
    return _security_gateway_instance

File: src/lib/security/SecurityGateway.ts
/**
 * SecurityGateway - Frontend Security Gateway
 * 
 * Provides input validation, sanitization, and Arabic error message translation
 * with graceful degradation for invalid inputs.
 * 
 * Week 2 Task 2.1: Security Implementation
 */

export interface ValidationResult {
  valid: boolean;
  sanitized?: any;
  error?: SecurityError;
}

export interface SecurityError {
  code: string;
  message: string;
  messageAr: string; // Arabic translation
  severity: 'info' | 'warning' | 'error' | 'critical';
  field?: string;
  details?: Record<string, any>;
}

export interface SecurityEvent {
  type: 'validation_failure' | 'sanitization' | 'suspicious_input' | 'rate_limit';
  timestamp: number;
  input?: any;
  error: SecurityError;
  userAgent?: string;
  url?: string;
}

/**
 * SecurityGateway - Main security validation and sanitization gateway
 */
export class SecurityGateway {
  private static instance: SecurityGateway;
  private eventLog: SecurityEvent[] = [];
  private readonly maxEventLogSize = 100;

  private constructor() {
    // Private constructor for singleton
  }

  static getInstance(): SecurityGateway {
    if (!SecurityGateway.instance) {
      SecurityGateway.instance = new SecurityGateway();
    }
    return SecurityGateway.instance;
  }

  /**
   * Validate and sanitize user input
   */
  validateInput<T>(input: T, rules?: ValidationRules): ValidationResult {
    try {
      // Basic type validation
      if (input === null || input === undefined) {
        return this.createErrorResult('INVALID_INPUT', 'Input is required', 'المدخل مطلوب');
      }

      // String validation
      if (typeof input === 'string') {
        return this.validateString(input, rules);
      }

      // Number validation
      if (typeof input === 'number') {
        return this.validateNumber(input, rules);
      }

      // Object validation
      if (typeof input === 'object' && !Array.isArray(input)) {
        return this.validateObject(input as Record<string, any>, rules);
      }

      // Array validation
      if (Array.isArray(input)) {
        return this.validateArray(input, rules);
      }

      return { valid: true, sanitized: input };
    } catch (error) {
      return this.createErrorResult(
        'VALIDATION_ERROR',
        `Validation failed: ${error instanceof Error ? error.message : 'Unknown error'}`,
        `فشل التحقق: ${error instanceof Error ? error.message : 'خطأ غير معروف'}`
      );
    }
  }

  /**
   * Validate string input
   */
  private validateString(input: string, rules?: ValidationRules): ValidationResult {
    // Check for dangerous patterns
    const dangerousPatterns = [
      /<script/i,
      /javascript:/i,
      /onerror=/i,
      /onload=/i,
      /eval\(/i,
      /expression\(/i,
      /data:text\/html/i,
    ];

    for (const pattern of dangerousPatterns) {
      if (pattern.test(input)) {
        this.logSecurityEvent('suspicious_input', {
          code: 'XSS_ATTEMPT',
          message: 'Input contains potentially dangerous content',
          messageAr: 'المدخل يحتوي على محتوى خطير محتمل',
          severity: 'error',
        }, input);
        return this.createErrorResult(
          'XSS_ATTEMPT',
          'Input contains potentially dangerous content',
          'المدخل يحتوي على محتوى خطير محتمل'
        );
      }
    }

    // SQL injection patterns
    const sqlPatterns = [
      /(\b(SELECT|INSERT|UPDATE|DELETE|DROP|CREATE|ALTER|EXEC|EXECUTE)\b)/i,
      /('|(\\')|(;)|(\\)|(\/\*)|(\*\/)|(--)|(#))/,
    ];

    for (const pattern of sqlPatterns) {
      if (pattern.test(input)) {
        this.logSecurityEvent('suspicious_input', {
          code: 'SQL_INJECTION_ATTEMPT',
          message: 'Input contains potentially dangerous SQL patterns',
          messageAr: 'المدخل يحتوي على أنماط SQL خطيرة محتملة',
          severity: 'error',
        }, input);
        return this.createErrorResult(
          'SQL_INJECTION_ATTEMPT',
          'Input contains potentially dangerous SQL patterns',
          'المدخل يحتوي على أنماط SQL خطيرة محتملة'
        );
      }
    }

    // Apply rules
    if (rules) {
      if (rules.minLength && input.length < rules.minLength) {
        return this.createErrorResult(
          'MIN_LENGTH',
          `Input must be at least ${rules.minLength} characters`,
          `يجب أن يكون المدخل ${rules.minLength} أحرف على الأقل`
        );
      }

      if (rules.maxLength && input.length > rules.maxLength) {
        return this.createErrorResult(
          'MAX_LENGTH',
          `Input must be no more than ${rules.maxLength} characters`,
          `يجب ألا يتجاوز المدخل ${rules.maxLength} حرف`
        );
      }

      if (rules.pattern && !rules.pattern.test(input)) {
        return this.createErrorResult(
          'PATTERN_MISMATCH',
          'Input does not match required pattern',
          'المدخل لا يطابق النمط المطلوب'
        );
      }
    }

    // Sanitize
    const sanitized = this.sanitizeString(input);
    
    if (sanitized !== input) {
      this.logSecurityEvent('sanitization', {
        code: 'INPUT_SANITIZED',
        message: 'Input was sanitized',
        messageAr: 'تم تنظيف المدخل',
        severity: 'info',
      }, input);
    }

    return { valid: true, sanitized };
  }

  /**
   * Validate number input
   */
  private validateNumber(input: number, rules?: ValidationRules): ValidationResult {
    if (isNaN(input) || !isFinite(input)) {
      return this.createErrorResult(
        'INVALID_NUMBER',
        'Input must be a valid number',
        'يجب أن يكون المدخل رقماً صحيحاً'
      );
    }

    if (rules) {
      if (rules.min !== undefined && input < rules.min) {
        return this.createErrorResult(
          'MIN_VALUE',
          `Input must be at least ${rules.min}`,
          `يجب أن يكون المدخل ${rules.min} على الأقل`
        );
      }

      if (rules.max !== undefined && input > rules.max) {
        return this.createErrorResult(
          'MAX_VALUE',
          `Input must be no more than ${rules.max}`,
          `يجب ألا يتجاوز المدخل ${rules.max}`
        );
      }
    }

    return { valid: true, sanitized: input };
  }

  /**
   * Validate object input
   */
  private validateObject(input: Record<string, any>, rules?: ValidationRules): ValidationResult {
    const sanitized: Record<string, any> = {};
    const errors: SecurityError[] = [];

    for (const [key, value] of Object.entries(input)) {
      const result = this.validateInput(value, rules);
      if (result.valid) {
        sanitized[key] = result.sanitized;
      } else if (result.error) {
        errors.push({ ...result.error, field: key });
      }
    }

    if (errors.length > 0) {
      return {
        valid: false,
        error: {
          code: 'OBJECT_VALIDATION_FAILED',
          message: 'Object validation failed',
          messageAr: 'فشل التحقق من الكائن',
          severity: 'error',
          details: { errors },
        },
      };
    }

    return { valid: true, sanitized };
  }

  /**
   * Validate array input
   */
  private validateArray(input: any[], rules?: ValidationRules): ValidationResult {
    const sanitized: any[] = [];
    const errors: SecurityError[] = [];

    for (let i = 0; i < input.length; i++) {
      const result = this.validateInput(input[i], rules);
      if (result.valid) {
        sanitized.push(result.sanitized);
      } else if (result.error) {
        errors.push({ ...result.error, field: `[${i}]` });
      }
    }

    if (errors.length > 0) {
      return {
        valid: false,
        error: {
          code: 'ARRAY_VALIDATION_FAILED',
          message: 'Array validation failed',
          messageAr: 'فشل التحقق من المصفوفة',
          severity: 'error',
          details: { errors },
        },
      };
    }

    return { valid: true, sanitized };
  }

  /**
   * Sanitize string input
   */
  private sanitizeString(input: string): string {
    // Remove dangerous characters
    let sanitized = input
      .replace(/[<>]/g, '') // Remove < and >
      .replace(/javascript:/gi, '')
      .replace(/on\w+=/gi, '') // Remove event handlers
      .trim();

    // HTML entity encoding for display
    sanitized = sanitized
      .replace(/&/g, '&amp;')
      .replace(/"/g, '&quot;')
      .replace(/'/g, '&#x27;');

    return sanitized;
  }

  /**
   * Create error result
   */
  private createErrorResult(
    code: string,
    message: string,
    messageAr: string,
    severity: SecurityError['severity'] = 'error'
  ): ValidationResult {
    return {
      valid: false,
      error: {
        code,
        message,
        messageAr,
        severity,
      },
    };
  }

  /**
   * Log security event
   */
  private logSecurityEvent(
    type: SecurityEvent['type'],
    error: SecurityError,
    input?: any
  ): void {
    const event: SecurityEvent = {
      type,
      timestamp: Date.now(),
      input: this.sanitizeForLogging(input),
      error,
      userAgent: typeof navigator !== 'undefined' ? navigator.userAgent : undefined,
      url: typeof window !== 'undefined' ? window.location.href : undefined,
    };

    this.eventLog.push(event);

    // Keep log size manageable
    if (this.eventLog.length > this.maxEventLogSize) {
      this.eventLog.shift();
    }

    // Send to backend if available
    this.sendSecurityEventToBackend(event).catch(() => {
      // Silently fail if backend is unavailable
    });
  }

  /**
   * Sanitize input for logging (remove sensitive data)
   */
  private sanitizeForLogging(input: any): any {
    if (typeof input === 'string') {
      // Truncate long strings
      if (input.length > 100) {
        return input.substring(0, 100) + '...';
      }
      return input;
    }
    return input;
  }

  /**
   * Send security event to backend
   */
  private async sendSecurityEventToBackend(event: SecurityEvent): Promise<void> {
    try {
      const apiBase = import.meta.env.VITE_API_URL || 'http://localhost:8000';
      await fetch(`${apiBase}/api/v2/security/events`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(event),
      });
    } catch (error) {
      // Silently fail - don't break user experience
      console.warn('Failed to send security event to backend:', error);
    }
  }

  /**
   * Get security event log (for debugging)
   */
  getEventLog(): SecurityEvent[] {
    return [...this.eventLog];
  }

  /**
   * Clear event log
   */
  clearEventLog(): void {
    this.eventLog = [];
  }

  /**
   * Get localized error message
   */
  getLocalizedError(
    errorCode: string,
    _lang: 'en' | 'ar' = 'en',
    details?: Record<string, any>
  ): SecurityError {
    const errorMessages: Record<string, { en: string; ar: string }> = {
      LOW_MEMORY_FALLBACK: {
        en: 'Low memory detected. Switching to 2D view for better performance.',
        ar: 'تم اكتشاف ذاكرة منخفضة. التبديل إلى عرض ثنائي الأبعاد لأداء أفضل.',
      },
      RENDERER_INITIALIZATION_FAILED: {
        en: 'Failed to initialize 3D renderer.',
        ar: 'فشل في تهيئة عارض ثلاثي الأبعاد.',
      },
      GEOMETRY_LOAD_FAILED: {
        en: 'Failed to load 3D geometry.',
        ar: 'فشل في تحميل الهندسة ثلاثية الأبعاد.',
      },
      WORKFLOW_STAGE_ERROR: {
        en: 'Workflow stage "{stage}" failed: {error}',
        ar: 'فشلت مرحلة "{stage}" من سير العمل: {error}',
      },
      WORKFLOW_ERROR: {
        en: 'Workflow error: {error}',
        ar: 'خطأ في سير العمل: {error}',
      },
      CNC_EXPORT_VALIDATION_FAILED: {
        en: 'CNC export validation failed: {errors}',
        ar: 'فشل التحقق من تصدير CNC: {errors}',
      },
      INTERNAL_ERROR: {
        en: 'An internal error occurred.',
        ar: 'حدث خطأ داخلي.',
      },
    };

    const messages = errorMessages[errorCode] || {
      en: `Error: ${errorCode}`,
      ar: `خطأ: ${errorCode}`,
    };

    return {
      code: errorCode,
      message: messages.en,
      messageAr: messages.ar,
      severity: 'warning',
      details,
    };
  }

  /**
   * Get localized message
   */
  getLocalizedMessage(
    messageKey: string,
    lang: 'en' | 'ar' = 'en',
    params?: Record<string, any>
  ): string {
    const messages: Record<string, { en: string; ar: string }> = {
      RENDERER_LOADING: {
        en: 'Loading 3D model...',
        ar: 'جارٍ تحميل النموذج ثلاثي الأبعاد...',
      },
      RENDERER_READY: {
        en: '3D model ready',
        ar: 'النموذج ثلاثي الأبعاد جاهز',
      },
      MEMORY_WARNING: {
        en: 'High memory usage detected',
        ar: 'تم اكتشاف استخدام ذاكرة مرتفع',
      },
    };

    const message = messages[messageKey] || {
      en: messageKey,
      ar: messageKey,
    };

    let text = lang === 'ar' ? message.ar : message.en;
    
    // Replace parameters
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        text = text.replace(`{${key}}`, String(value));
      });
    }

    return text;
  }

  /**
   * Log security event (public method for external use)
   */
  logSecurityEventPublic(
    type: SecurityEvent['type'],
    severity: SecurityError['severity'],
    details?: Record<string, any>
  ): void {
    const error: SecurityError = {
      code: type.toUpperCase(),
      message: `Security event: ${type}`,
      messageAr: `حدث أمني: ${type}`,
      severity,
      details,
    };

    this.logSecurityEvent(type, error);
  }
}

/**
 * Validation rules interface
 */
export interface ValidationRules {
  minLength?: number;
  maxLength?: number;
  min?: number;
  max?: number;
  pattern?: RegExp;
  required?: boolean;
}

/**
 * Export singleton instance
 */
export const securityGateway = SecurityGateway.getInstance();

/**
 * Convenience functions
 */
export function validateInput<T>(input: T, rules?: ValidationRules): ValidationResult {
  return securityGateway.validateInput(input, rules);
}

export function sanitizeString(input: string): string {
  const result = securityGateway.validateInput(input);
  return result.valid && typeof result.sanitized === 'string' 
    ? result.sanitized 
    : '';
}

